src/main/src/main/java/org/geoserver/security/impl/LayerGroupContainmentCache.java

```diff
@@ -0,0 +1,426 @@
+     */
+    Map<String, LayerGroupSummary> groupCache = new ConcurrentHashMap<>();
+
+    /**
+     * Lookup from {@link ResourceInfo} id to groups directly containing its associated layers (the transitive containment is computed by suing
+     * {@link LayerGroupSummary}
+     */
+    Map<String, Set<LayerGroupSummary>> resourceContainmentCache = new ConcurrentHashMap<>();
+
+    private Catalog catalog;
+
+    public LayerGroupContainmentCache(Catalog catalog) {
+        this.catalog = catalog;
+        catalog.addListener(new CatalogChangeListener());
+        buildLayerGroupCaches();
+    }
+
+    private void buildLayerGroupCaches() {
+        groupCache.clear();
+        resourceContainmentCache.clear();
+        List<LayerGroupInfo> groups = catalog.getLayerGroups();
+
+        // first populate the basic structure
+        for (LayerGroupInfo lg : groups) {
+            addGroupInfo(lg);
+        }
+
+        // now populate the containment structure
+        for (LayerGroupInfo lg : groups) {
+            registerContainedGroups(lg);
+        }
+
+    }
+
+    private void registerContainedGroups(LayerGroupInfo lg) {
+        lg.getLayers().stream().filter(IS_GROUP).forEach(p -> {
+            String containerId = lg.getId();
+            String containedId = p.getId();
+            LayerGroupSummary container = groupCache.get(containerId);
+            LayerGroupSummary contained = groupCache.get(containedId);
+            if (container != null && contained != null) {
+                contained.containerGroups.add(container);
+            }
+        });
+    }
+
+    private void addGroupInfo(LayerGroupInfo lg) {
+        LayerGroupSummary groupData = new LayerGroupSummary(lg);
+        groupCache.put(lg.getId(), groupData);
+        lg.getLayers().stream().filter(IS_LAYER).forEach(p -> {
+            String id = ((LayerInfo) p).getResource().getId();
+            Set<LayerGroupSummary> containers = resourceContainmentCache.computeIfAbsent(id,
+                    CONCURRENT_SET_BUILDER);
+            containers.add(groupData);
+        });
+    }
+
+    private void clearGroupInfo(LayerGroupInfo lg) {
+        LayerGroupSummary data = groupCache.remove(lg.getId());
+        // clear the resource containment cache
+        lg.getLayers().stream().filter(IS_LAYER).forEach(p -> {
+            String rid = ((LayerInfo) p).getResource().getId();
+            synchronized (rid) {
+                Set<LayerGroupSummary> containers = resourceContainmentCache.get(rid);
+                if (containers != null) {
+                    containers.remove(data);
+                }
+            }
+        });
+        // this group does not contain anything anymore, remove from containment
+        for (LayerGroupSummary d : groupCache.values()) {
+            d.containerGroups.remove(lg.getId());
+        }
+    }
+
+    /**
+     * Returns all groups containing directly or indirectly containing the resource
```

Comment:
Returns all groups **containing** directly or indirectly **containing** the resource